log-replication/src/main/java/org/corfudb/logreplication/fsm/LogReplicationFSM.java

```diff
@@ -0,0 +1,352 @@
+ *
+ *
+ *                                       replication_stop
+ *                      +-------------------------------------------------+
+ *    replication_stop  |                                                 |
+ *             +-----+  |              replication_stop                   |
+ *             |     |  v      v-----------------------------+            |
+ *             |    ++--+---------+                          |        +---+--------------------+
+ *             +--->+ INITIALIZED +------------------------+ |        | SNAPSHOT_SYNC_REQUIRED +<---+
+ *                  +---+----+----+ snapshot_sync_request  | |        +---+---------------+----+    |
+ *                      ^    |                             | |            |               ^         |
+ *                      |    |                             | |   snapshot |               |         |
+ *                      |    |                             | |    sync    |               |         |
+ *     replication_stop |    | replication_start           | |    request |               |         |
+ *                      |    |                             | |            |               |         |
+ *                      |    v                             v |            v               |         |
+ *               +------+----+-------+  snapshot_sync    +-+-+------------+-+             |         |
+ *         +-----| IN_LOG_ENTRY_SYNC |     request       | IN_SNAPSHOT_SYNC +             |         |
+ *         |     |                   +------------------>+                  |             |         |
+ *         |     +----+----+---------+                   +---+---+----------+-------------+         |
+ *         |       ^  |   ^                                 |    |        ^        sync             |
+ *         |       |  |   +---------------------------------+    |        |       cancel            |
+ *         + ----- +  |                snapshot_sync             + -------+                         |
+ *  log_entry_sync    |                  complete               snapshot_sync                       |
+ *    continue        |                                           continue                          |
+ *                    +-----------------------------------------------------------------------------+
+ *                                                     sync_cancel
+ *               replication
+ * +---------+    shutdown    +------------+
+ * | STOPPED +<---------------+ ALL_STATES |
+ * +---------+                +------------+
+ *
+ *
+ */
+@Slf4j
+public class LogReplicationFSM {
+
+    @Getter
+    private long siteConfigID;
+
+    /**
+     * Current state of the FSM.
+     */
+    @Getter
+    private volatile LogReplicationState state;
+
+    /**
+     * Map of all Log Replication FSM States (reuse single instance for each state)
+     */
+    @Getter
+    private Map<LogReplicationStateType, LogReplicationState> states = new HashMap<>();
```

Comment:
instead of such a map linking the state types and state objects, I suggest having private final stateType variable inside the state objects.  Because the state type is a property of the object, not determined by the map created here